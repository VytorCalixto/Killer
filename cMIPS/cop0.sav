[*]
[*] GTKWave Analyzer v3.3.37 (w)1999-2012 BSI
[*] Fri Mar 20 20:27:42 2015
[*]
[dumpfile] "/home/roberto/cMIPS/v_cMIPS.vcd"
[dumpfile_mtime] "Fri Mar 20 19:54:48 2015"
[dumpfile_size] 21630696
[savefile] "/home/roberto/cMIPS/cop0.sav"
[timestart] 5829800000
[size] 1062 920
[pos] 854 0
*-26.000000 5940000000 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1
[treeopen] u_core.
[treeopen] u_core.u_alu.
[sst_width] 210
[signals_width] 265
[sst_expanded] 1
[sst_vpaned_height] 291
@28
clk
@200
-    fetch
@22
u_core.pc[31:0]
u_core.instr_fetched[31:0]
@28
u_core.pcsel[1:0]
u_core.excp_pcsel[2:0]
u_core.exception_stall
@24
u_core.nullify
@29
u_core.interrupt_taken
@200
-    decode, reg fetch
@22
u_core.rf_instruction[31:0]
@24
u_core.a_rs[4:0]
u_core.a_rt[4:0]
u_core.a_rd[4:0]
u_core.exception_dec[31:0]
@200
-    exec
@24
u_core.u_alu.operation[31:0]
@22
u_core.alu_inp_a[31:0]
u_core.alu_inp_b[31:0]
u_core.result[31:0]
@28
u_core.ovfl
@200
-   COP-0
@24
u_core.exccode[4:0]
u_core.exception_num[31:0]
@22
u_core.cause[31:0]
u_core.status[31:0]
u_core.epc[31:0]
@200
-  counter
@28
u_core.irq[5:0]
@22
u_interrupt_counter.q[29:0]
@200
-    memory
@28
u_core.b_sel[3:0]
@22
u_core.data_inp[31:0]
u_core.data_out[31:0]
d_addr[31:0]
@28
cpu_d_aval
wr
@200
-    write-back
@28
u_core.wb_wreg
@24
u_core.wb_a_c[4:0]
@22
u_core.wb_cop0_val[31:0]
@28
u_core.wb_muxc[2:0]
@22
u_core.wb_c[31:0]
[pattern_trace] 1
[pattern_trace] 0
