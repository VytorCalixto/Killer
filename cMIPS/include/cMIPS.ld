SECTIONS
{
       . = 0x00000000; /* x_INST_BASE_ADDR */
       .text : { *(.text .text.*) }
       . = 0x00040000; /* x_DATA_BASE_ADDR */
       .data    : { *(.data .data.*) }
       .rodata  : { *(.rodata .rodata.*) }
       .rodata1 : { *(.rodata1) }
       .data1   : { *(.data1) }
       .sdata   : { *(.sdata .sdata.*) }
       .lit8    : { *(.lit8) }
       .lit4    : { *(.lit4) }
       .sbss    : { *(.sbss .sbss.*) *(.scommon .scommon.*) }
       .bss     : { *(.bss .bss.*) *(COMMON) }
}

