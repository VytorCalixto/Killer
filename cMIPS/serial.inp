0
11
ffffffff
7C2



