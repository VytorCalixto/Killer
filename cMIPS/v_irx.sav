[*]
[*] GTKWave Analyzer v3.3.37 (w)1999-2012 BSI
[*] Sun May 17 23:47:43 2015
[*]
[dumpfile] "/home/roberto/cMIPS/v_cMIPS.vcd"
[dumpfile_mtime] "Sun May 17 23:34:01 2015"
[dumpfile_size] 24210019
[savefile] "/home/roberto/cMIPS/v_irx.sav"
[timestart] 6435000000
[size] 1200 908
[pos] 676 118
*-29.000000 7800000000 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1
[treeopen] u_simple_uart.
[sst_width] 210
[signals_width] 221
[sst_expanded] 1
[sst_vpaned_height] 266
@28
clk
@22
u_core.pc[31:0]
@200
-    decode, reg fetch
@22
u_core.rf_instruction[31:0]
u_core.regs_a[31:0]
u_core.regs_b[31:0]
@200
-    exec
@24
u_core.u_alu.operation[31:0]
@22
u_core.alu_inp_a[31:0]
u_core.alu_inp_b[31:0]
u_core.result[31:0]
@200
-  COP0
@28
u_core.interrupt_taken
@22
u_core.epc[31:0]
u_core.cause[31:0]
u_core.status[31:0]
@200
-    memory
@28
u_core.b_sel[3:0]
@22
d_addr[31:0]
u_core.data_inp[31:0]
u_core.data_out[31:0]
@28
cpu_d_aval
u_core.mm_wrmem
@200
-    write-back
@28
u_core.wb_muxc[2:0]
u_core.wb_wreg
@22
u_core.wb_a_c[4:0]
u_core.wb_c[31:0]
@200
-  UART
@28
u_simple_uart.u_uart.s_stat
@22
u_simple_uart.u_uart.status[7:0]
@28
u_simple_uart.u_uart.s_ctrl
@22
u_simple_uart.u_uart.ctrl[7:0]
@28
u_simple_uart.u_uart.s_rx
@22
u_simple_uart.u_uart.rxreg[7:0]
@28
u_simple_uart.irq
@200
-  reception
@24
u_simple_uart.u_uart.rxcpu_dbg_st[31:0]
@28
u_simple_uart.u_uart.rx_bfr_full
@24
u_simple_uart.u_uart.rx_dbg_st[31:0]
@28
u_simple_uart.u_uart.rxclk
u_simple_uart.u_uart.rxdat
@200
-  REMOTE (fake) UART
@24
u_uart_remota.tx_dbg_st[31:0]
@28
u_uart_remota.outdat
[pattern_trace] 1
[pattern_trace] 0
