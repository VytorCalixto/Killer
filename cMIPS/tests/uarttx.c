
#include "cMIPS.h"

typedef struct control { // control register fields (uses only ls byte)
  int ign   : 24,        // ignore uppermost bits
    rts     : 1,         // Request to Send
    ign2    : 2,         // bits 6,5 ignored
    intTX   : 1,         // interrupt on TX buffer empty (bit 4)
    intRX   : 1,         // interrupt on RX buffer full (bit 3)
    speed   : 3;         // 4,8,16..256 tx-rx clock data rates  (bits 0..2)
} Tcontrol;

typedef struct status { // status register fields (uses only ls byte)
  int s;
  // int ign   : 24,      // ignore uppermost bits
  //  ign7    : 1,        // ignored (bit 7)
  //  txEmpty : 1,        // TX register is empty (bit 6)
  //  rxFull  : 1,        // octet available from RX register (bit 5)
  //  int_TX_empt: 1,     // interrupt pending on TX empty (bit 4)
  //  int_RX_full: 1,     // interrupt pending on RX full (bit 3)
  //  ign2    : 1,        // ignored (bit 2)
  //  framing : 1,        // framing error (bit 1)
  //  overun  : 1;        // overun error (bit 0)
} Tstatus;

#define RXfull  0x00000020
#define TXempty 0x00000040

typedef union ctlStat { // control + status on same address
  Tcontrol  ctl;        // write-only
  Tstatus   stat;       // read-only
} TctlStat;

typedef union data {    // data registers on same address
  int tx;               // write-only
  int rx;               // read-only
} Tdata;

typedef struct serial {
  TctlStat cs;
  Tdata    d;
} Tserial;

#define LONG_STRING 1

#if LONG_STRING
char *dog = "\n\tthe quick brown fox jumps over the lazy dog\n";
char s[32];
#else
char s[32]; //  = "123";
#endif

int strcopy(const char *y, char *x)
{
  int i=0;
  while ( (*x++ = *y++) != '\0' ) // copia e testa final
    i = i+1;
  *x = '\0';
  return(i+1);
}



#define SPEED 0       // operate at the highest data rate

#define COUNTING ((SPEED+1)*100) // how long to wait for last bits to be sent out


int main(void) { // send a string through the UART serial interface
  int i;
  volatile int state, val;
  volatile Tserial *uart;  // tell GCC to not optimize away tests
  Tcontrol ctrl;

  volatile int *counter;        // address of counter

#if LONG_STRING
  i = strcopy(dog, s);
#else
  s[0] = '1';   s[1] = '2';   s[2] = '3';   s[3] = '\0';
#endif 

  uart = (void *)IO_UART_ADDR;  // UART's address

  counter = (int *)IO_COUNT_ADDR; // counter's address

  ctrl.speed = SPEED;
  ctrl.intTX = 0;  // no interrupts
  ctrl.intRX = 0;
  ctrl.ign2  = 0;
  ctrl.ign   = 0;
  ctrl.rts   = 1;
  uart->cs.ctl = ctrl;

  i = -1;
  do {

    i = i+1;
    // while ( (state = uart->cs.stat.txEmpty) != 1 )
    while ( ! ( (state = uart->cs.stat.s) & TXempty ) )
      ;
    uart->d.tx = (int)s[i];

  } while (s[i] != '\0');  // '\0' is transmitted in previous line


  // then wait until last char is sent out of the shift-register to return
  startCounter(COUNTING, 0);

  while ( (val=(readCounter() & 0x3fffffff)) < COUNTING )
    {}; 

  return val;  // so compiler won't optimize away the last loop
}
