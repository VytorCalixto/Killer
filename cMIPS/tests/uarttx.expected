1
2
3
00000006
00000016
